internal/provider/vm/vm_affinity_rule_resource.go

```diff
@@ -0,0 +1,437 @@
+				foundRules = append(foundRules, rule)
+				break
+			}
+		} else {
+			if rule.Name == affinityRuleIdentifier {
+				foundRules = append(foundRules, rule)
+			}
+		}
+	}
+
+	if len(foundRules) == 0 {
+		resp.Diagnostics.AddError("Failed to import resource", "no affinity rule found with name or ID "+affinityRuleIdentifier)
+		return
+	}
+	if len(foundRules) > 1 {
+		resp.Diagnostics.AddError("Failed to import resource", "more than one affinity rule found with name or ID "+affinityRuleIdentifier)
+		return
+	}
+	vmAffinityRule := foundRules[0]
+	state = &vmAffinityRuleResourceModel{
+		ID:   types.StringValue(vmAffinityRule.ID),
+		Name: types.StringValue(vmAffinityRule.Name),
+		VDC:  types.StringValue(vdcName),
+	}
+
+	endpointVMs := vmReferencesToListValue(vmAffinityRule.VmReferences)
+	state.VMIDs = types.ListValueMust(types.StringType, endpointVMs)
+
+	// Set state to fully populated data
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// resourceToAffinityRule prepares a VM affinity rule definition from the data in the resource.
+func resourceToAffinityRule(r *vmAffinityRuleResource, m *vmAffinityRuleResourceModel) (*govcdtypes.VmAffinityRule, error) {
+	name := m.Name.ValueString()
+	polarity := m.Polarity.ValueString()
+	required := m.Required.ValueBool()
+	enabled := m.Enabled.ValueBool()
+	rawVms := m.VMIDs
+
+	fullVMList, err := r.vdc.QueryVmList(govcdtypes.VmQueryFilterOnlyDeployed)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		vmReferences   []*govcdtypes.Reference
+		invalidEntries = make(map[string]bool)
+		foundEntries   = make(map[string]bool)
+	)
+
+	for _, vmID := range rawVms.Elements() {
+		for _, vm := range fullVMList {
+			uuid := common.ExtractUUID(vmID.String())
+			if uuid != "" {
+				if uuid == common.ExtractUUID(vm.HREF) {
+					vmReferences = append(vmReferences, &govcdtypes.Reference{HREF: vm.HREF})
+					foundEntries[vmID.String()] = true
+				}
+			} else {
+				invalidEntries[vmID.String()] = true
+			}
+		}
+	}
+	if len(invalidEntries) > 0 {
+		var invalidItems []string
+		for k := range invalidEntries {
+			invalidItems = append(invalidItems, k)
+		}
+		return nil, fmt.Errorf("invalid entries (not a VM ID) detected: %v", invalidItems)
+	}
+	if len(rawVms.Elements()) > len(foundEntries) {
+		var notExistingVms []string
+		for _, vmID := range rawVms.Elements() {
+			_, exists := foundEntries[vmID.String()]
+			if !exists {
+				notExistingVms = append(notExistingVms, vmID.String())
+			}
+		}
+		return nil, fmt.Errorf("not existing VMs detected: %v", notExistingVms)
+	}
+
+	vmAffinityRuleDef := &govcdtypes.VmAffinityRule{
+		Name:        name,
+		IsEnabled:   &enabled,
+		IsMandatory: &required,
+		Polarity:    polarity,
+		VmReferences: []*govcdtypes.VMs{
+			{
+				VMReference: vmReferences,
+			},
+		},
+	}
+	return vmAffinityRuleDef, nil
+}
+
+func getVMAffinityRule(vdc vdc.VDC, name, id string) (*govcd.VmAffinityRule, error) {
```

Comment:
it is better to extend the global struct in order to be able to access the resources (ex: vdc)

```suggestion
func (r *vmAffinityRuleResource) getVMAffinityRule(name, id string) (*govcd.VmAffinityRule, error) {
```

Reply:
this function is used by datasource